File: Laravel/app/Http/Controllers/BlogController.php
<?php

namespace App\Http\Controllers;

use App\Bd\Bd;
use App\Bd\Client;
use Illuminate\View\View;

class BlogController extends Controller
{
    public function index(): View
    {
        $posts = Bd::remember(
            'blog:list',
            ['bd:blog'],
            static fn (Client $c) => $c->blog()->listPosts(limit: 20),
            default: ['items' => []],
        );

        return view('pages.blog', ['posts' => $posts['items'] ?? []]);
    }

    public function show(string $slug): View
    {
        $post = Bd::remember(
            "blog:post:{$slug}",
            ['bd:blog', "bd:blog:{$slug}"],
            static fn (Client $c) => $c->blog()->getPost($slug),
            default: null,
        );

        abort_if($post === null, 404);

        $comments = Bd::remember(
            "blog:comments:{$slug}",
            ['bd:blog', "bd:blog:{$slug}"],
            static fn (Client $c) => $c->blog()->listComments($slug, limit: 50),
            default: ['items' => []],
        );

        // The API wraps the post: `{ post, access }` — `access` is
        // "granted" or "paywall", which is how a paywalled post is signalled.
        return view('pages.blog-post', [
            'post' => $post['post'] ?? [],
            'access' => $post['access'] ?? 'granted',
            'comments' => $comments['items'] ?? [],
        ]);
    }
}

File: Phoenix/lib/bd_starter_web/controllers/blog_controller.ex
defmodule BdStarterWeb.BlogController do
  use BdStarterWeb, :controller

  alias BdStarter.Bd

  def index(conn, _params), do: render(conn, :index, posts: Bd.posts(20))

  def show(conn, %{"slug" => slug}) do
    # The API wraps the post: `%{"post" => …, "access" => "granted" | "paywall"}`.
    # `access` is how a paywalled post is signalled — the body comes back
    # truncated rather than absent, so a template that ignores it silently
    # renders a teaser as if it were the whole article.
    case Bd.post(slug) do
      nil ->
        conn |> put_status(:not_found) |> text("Post not found")

      %{"post" => post} = wrapper ->
        render(conn, :show,
          post: post,
          access: Map.get(wrapper, "access", "granted"),
          comments: Bd.post_comments(slug)
        )

      _ ->
        conn |> put_status(:not_found) |> text("Post not found")
    end
  end
end
